src/RestApi/StoreApi/Controllers/CartOrder.php

```diff
@@ -0,0 +1,362 @@
+use \WP_REST_Request as RestRequest;
+use \WC_REST_Exception as RestException;
+use Automattic\WooCommerce\Blocks\RestApi\StoreApi\Schemas\OrderSchema;
+
+/**
+ * Cart Order API.
+ *
+ * Creates orders based on cart contents.
+ */
+class CartOrder extends RestController {
+	/**
+	 * Endpoint namespace.
+	 *
+	 * @var string
+	 */
+	protected $namespace = 'wc/store';
+
+	/**
+	 * Route base.
+	 *
+	 * @var string
+	 */
+	protected $rest_base = 'cart/order';
+
+	/**
+	 * Schema class instance.
+	 *
+	 * @var object
+	 */
+	protected $schema;
+
+	/**
+	 * Setup API class.
+	 */
+	public function __construct() {
+		$this->schema = new OrderSchema();
+	}
+
+	/**
+	 * Register routes.
+	 */
+	public function register_routes() {
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base,
+			[
+				[
+					'methods'  => RestServer::CREATABLE,
+					'callback' => array( $this, 'create_item' ),
+					'args'     => $this->get_endpoint_args_for_item_schema( RestServer::CREATABLE ),
+				],
+				'schema' => [ $this, 'get_public_item_schema' ],
+			]
+		);
+	}
+
+	/**
+	 * Converts the global cart to an order object.
+	 *
+	 * @todo Since this relies on the cart global so much, why doesn't the core cart class do this?
+	 * @todo set payment method
+	 * @todo set customer note
+	 *
+	 * Based on WC_Checkout::create_order.
+	 *
+	 * @param RestRequest $request Full details about the request.
+	 * @return RestError|RestResponse
+	 */
+	public function create_item( $request ) {
+		add_filter( 'woocommerce_default_order_status', array( $this, 'default_order_status' ) );
+
+		try {
+			// Create or retrieve the draft order for the current cart.
+			$order_object  = $this->create_order_from_cart( $request );
+			$reserve_stock = $this->reserve_stock_for_draft_order( $order_object );
+
+			if ( is_wp_error( $reserve_stock ) ) {
+				// Something went wrong - return error.
+				$response = $reserve_stock;
+			} else {
+				$response = $this->prepare_item_for_response( $order_object, $request );
+				$response->set_status( 201 );
+			}
+		} catch ( Exception $e ) {
+			$response = new RestError( 'checkout-error', $e->getMessage() );
+		}
+
+		remove_filter( 'woocommerce_default_order_status', array( $this, 'default_order_status' ) );
+
+		return $response;
+	}
+
+	/**
+	 * Put a temporary hold on stock for this order.
+	 *
+	 * @throws RestException Exception when stock cannot be reserved.
+	 * @param \WC_Order $order Order object.
+	 * @return bool|RestError
+	 */
+	protected function reserve_stock_for_draft_order( $order ) {
+		global $wpdb;
+
+		try {
+			$hold_stock_minutes = (int) get_option( 'woocommerce_hold_stock_minutes', 0 );
+			$stock_to_reserve   = [];
+
+			// Loop over line items and check each item may be purchased.
+			foreach ( $order->get_items() as $item ) {
+				if ( ! $item->is_type( 'line_item' ) ) {
+					continue;
+				}
+
+				$product = $item->get_product();
+
+				if ( ! $product->is_in_stock() ) {
```

Comment:
Should check `$product` is an instance of `\WC_Product` before calling a method on it.